metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py

```diff
@@ -0,0 +1,179 @@
+)
+from datahub.emitter.mcp import MetadataChangeProposalWrapper
+from datahub.ingestion.api.source import Source, SourceReport
+from datahub.ingestion.api.workunit import MetadataWorkUnit, UsageStatsWorkUnit
+
+logger = logging.getLogger(__name__)
+
+auditStamp = models.AuditStampClass(
+    time=get_sys_time(), actor="urn:li:corpUser:pythonEmitter"
+)
+
+
+class EntityConfig(ConfigModel):
+    name: str
+    type: str
+    env: str
+    platform: str
+    platform_instance: Optional[str]
+
+
+class EntityNodeConfig(ConfigModel):
+    entity: EntityConfig
+    upstream: Optional[List["EntityNodeConfig"]]
+
+
+# https://pydantic-docs.helpmanual.io/usage/postponed_annotations/ required for when you reference a model within itself
+EntityNodeConfig.update_forward_refs()
+
+
+class LineageFileSourceConfig(ConfigModel):
+    file: str
+    preserve_upstream: bool = True
+
+
+class DefaultConfig(ConfigModel):
+    version: str = "1"
+
+    @validator("version")
+    def version_must_be_1(cls, v):
+        if v != "1":
+            raise ValueError("Only version 1 is supported")
+
+
+class LineageConfig(DefaultConfig):
+    lineage: List[EntityNodeConfig]
+
+
+@dataclass
+class LineageFileSource(Source):
+    config: LineageFileSourceConfig
+    report: SourceReport = field(default_factory=SourceReport)
+
+    @classmethod
+    def create(cls, config_dict, ctx):
+        config = LineageFileSourceConfig.parse_obj(config_dict)
+        return cls(ctx, config)
+
+    @staticmethod
+    def load_lineage_config(file_name: str) -> LineageConfig:
+        config = load_config_file(file_name)
+        lineage_config = LineageConfig.parse_obj(config)
+        return lineage_config
+
+    @staticmethod
+    def get_lineage_metadata_change_event_proposal(
+        entities: List[EntityNodeConfig], preserve_upstream: bool
+    ) -> Iterable[MetadataChangeProposalWrapper]:
+        """
+        Builds a list of events to be emitted to datahub by going through each entity and its upstream nodes
+        :param preserve_upstream: This field determines if we want to query the datahub backend to extract
+        the existing upstream lineages for each entity and preserve it
+        :param entities: A list of entities we want to build a proposal on
+        :return: Returns a list of metadata change event proposals to be emitted to datahub
+        """
+
+        def _get_entity_urn(entity_config: EntityConfig) -> Optional[str]:
+            """Helper inner function to extract a given entity_urn
+            A return value of None represents an unsupported entity type
+            """
+            if entity_config.type == "dataset":
+                return make_dataset_urn_with_platform_instance(
+                    platform=entity_config.platform,
+                    name=entity_config.name,
+                    env=entity_config.env,
+                    platform_instance=entity_config.platform_instance,
+                )
+            return None
```

Comment:
Log a warning if an unsupported type is encountered?

Reply:
Handled in other locations, but i suppose i could place a log in here as well.

https://github.com/linkedin/datahub/blob/a48884340263af12de9088c3fba079ed3fabfde9/metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py#L136-L140

https://github.com/linkedin/datahub/blob/a48884340263af12de9088c3fba079ed3fabfde9/metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py#L151-L155